### PR TITLE
Disconnect received videos when not needed

### LINKED ISSUE #5388: personally block the video of others to reduce internet use
When we go to disable other people's cameras for ourselves it would be useful not to receive the video stream of other people to reduce the use of internet bandwidth. Instead, now if you turn off other people's videos, we don't see their video streaming but it continues to get there.

### PR DESCRIPTION
**Pull request by danxuliu**
Fixes #5388
Requires https://github.com/strukturag/nextcloud-spreed-signaling/pull/191

If a received video is explicitly disabled by the user, or if the remote participant is not visible, the received video is not needed. Until now the video was simply not rendered, but with this pull request the received video is disconnected/blocked, so it is not even transmitted nor decoded.

Note that, for now, this works only when the HPB is used. When the HPB is not used the renegotiation would need to be implemented by all clients, so something for the future.

If the external signaling server does not support the new `update-sdp` feature the video will not be disconnected when not needed either. If `update-sdp` is not supported then requesting a new offer would close the previous connection and create a new one, which would cause the audio to be interrupted during the connection change.

Pending:
- [X] Disconnect also not visible videos (not only explicitly blocked videos)
- [X] Verify that overlapping disconnections and reconnections work as expected (for example, quickly blocking and unblocking again before the previous reconnection finished) - The RTCPeerConnection will trigger a new `negotiationneeded` event if the expected transceiver direction does not match once the connection is established
- [X] Adjust to possible changes in https://github.com/strukturag/nextcloud-spreed-signaling/pull/191
- [ ] Add details to the external signaling documentation about offer updates - The documentation was moved to the _nextcloud-spreed-signaling_ repository, so it should be done there
- [ ] Disconnect received video if the sender disabled it? Probably not worth the trouble, as the received video will be a 1FPS black stream, so it should not require much resources, and it would be needed only for the visible participants, as for the hidden ones it will be already blocked. Nevertheless, it could be added in a follow up if needed.
- [ ] Make this work also without HPB - Requires renegotiations in all the clients, it will be done in a follow up
- [ ] Show loading spinner or something like that while enabling video again? Or just the avatar? Right now the last video frame received before the video was disconnected may be shown even a few seconds after the renegotiation has finished. 
  - The `framesDecoded` property of the `inbound-rtp` stat could be monitored to know when the video has started to play again. However, in older browser versions that property was not available, and stats for received connections were not updated at a regular rate, so this should be carefully tested. Therefore it would be better to do it in a follow up pull request.
  - If instead of a WebRTC renegotiation the received video is disabled using the Janus specific approach described in https://github.com/strukturag/nextcloud-spreed-signaling/pull/191 the video seems to start playing again much faster, so it may not be needed to show a spinner. However, it would be still needed once the videos are disconnected also without HPB.

## How to test (scenario 1)

- Setup the HPB (with strukturag/nextcloud-spreed-signaling#191)
- Start a call with audio and video
- In a private window, join the call without audio nor video
- Open the WebRTC stats in another tab; in Firefox `about:webrtc`, in Chromium `chrome://webrtc-internals`
- Open the video stats for the receiver connection; in Firefox the stats are shown at the end of the details of a connection, in Chromium they are shown under `RTCMediaStreamTrack_receiver_2 (track)` (or something similar)
- Disable the received video

### Result with this pull request

The video is no longer received according to the stats; if the video is enabled again the video will be received again

### Result without this pull request

The video is still received according to the stats

## How to test (scenario 2)

- Setup the HPB (with strukturag/nextcloud-spreed-signaling#191)
- Start a call with audio and video
- In a private window, join the call without audio nor video
- In another browser, join the call without audio nor video
- Open the WebRTC stats in another tab; in Firefox `about:webrtc`, in Chromium `chrome://webrtc-internals`
- Open the video stats for the receiver connection; in Firefox the stats are shown at the end of the details of a connection, in Chromium they are shown under `RTCMediaStreamTrack_receiver_2 (track)` (or something similar)
- Follow the participant from the private window (so the participant from the original window is hidden)
- Wait 5 seconds (as the video is not immediately blocked when hidden)

### Result with this pull request

The video is no longer received according to the stats; if the video is shown again the video will be received again

### Result without this pull request

The video is still received according to the stats

### CODE REVIEW

File: src/__mocks__/attachmediastream.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Basic "attachmediastream" implementation without using "webrtc-adapter", as
+// "browserDetails" is null in unit tests.
+export default function(stream, element, options) {
```
Comment (nickvergessen) [inline src/__mocks__/attachmediastream.js:3]: GitHub Actions / eslint

src/__mocks__/attachmediastream.js#L3

**Comment by nickvergessen** [comment]
> In a private window, join the call without audio nor video
> The video is no longer received according to the stats; if the video is enabled again the video will be received again

So it worked for me using a private window, but a normal window did not work 😕 . Is that a bug with about:webrtc?

**Review by nickvergessen — APPROVED** [review]
Other than that it  seems to work.
PR state is still developing?

**Comment by danxuliu** [comment]
> > In a private window, join the call without audio nor video
> > The video is no longer received according to the stats; if the video is enabled again the video will be received again
> 
> So it worked for me using a private window, but a normal window did not work confused . Is that a bug with about:webrtc?

No idea :-/ I have just tried and it worked fine for me when blocking the video in a normal window and when blocking the video in a private window :shrug:

> Other than that it seems to work.
> PR state is still developing?

The pull request is ready, but it may require changes if something is changed in https://github.com/strukturag/nextcloud-spreed-signaling/pull/191 (that is why I did not set it to _To review_).

**Comment by danxuliu** [comment]
https://github.com/strukturag/nextcloud-spreed-signaling/pull/191 was merged, so ready for review. The pending items can be done in follow ups.

**Pull request merged**